### PR TITLE
Fixed issue #15 Upgrade to Asciidoctor 2.2.0

### LINKED ISSUE #4: Embed CSS to generated html file
Which option can I set to embed asciidoctor css to my generated html ? 
I try to set linkcss to true, but without success.

I also try to find an option to copy the asciidoctor css in the same destination folder but it still to be "./asciidoctor.css"

Here is my code

``` javascript
gulp.src('doc.adoc')
        .pipe(asciidoctor({
            safe: false,
            linkcss: false
        }))
        .pipe(gulp.dest('target/doc/'));
```

### LINKED ISSUE #11: Backend and templates
I want to use gulp-asciidoctor to render my adoc files to reveal.js slides
asciidoctor-reveal.js has usable custom templates
I've been wondering how to use it

any idea?

### LINKED ISSUE #15: Upgrade to Asciidoctor.js 2.2.0 and support templates
- Upgrade the gulp plugin to Asciidoctor.js version 2.2.0.
- Only add dependency to @asciidoctor/core
- Add support the new standalone option
- Add support for templates introduced in Asciidoctor 2.2.0
- Add testcases for templates
- Enhance documentation to show how templates can be used.

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,23 +5,25 @@
   "license": "MIT",
   "repository": "https://github.com/asciidoctor/gulp-asciidoctor",
   "author": {
-    "name": "John Dong",
-    "email": "dongwqs@sina.com",
-    "url": "https://github.com/dongwq/gulp-asciidoctor"
+    "name": "Armin Pfarr",
```
Comment (ggrossetie) [inline package.json]: I think you should add multiples authors:

```json
"authors": [
  "John Dong <dongwqs@sina.com>",
  "Armin Pfarr <arp@henriette-einstein.com>"
]
```

The original author seems to be https://github.com/jeremy889988 but he was never included in the package.json

File: doc/documentation.adoc [hunk 2]
```diff
@@ -6,7 +6,7 @@
 [source,javascript]
 ----
 gulp.pipe(asciidoctor({
-  header_footer: false,
+  header_footer: true,
```
Comment (ggrossetie) [inline doc/documentation.adoc]: We should probably use the new terminology in the example to encourage users to use it:

```suggestion
  standalone: true,
```

File: doc/documentation.adoc [hunk 3]
```diff
@@ -24,6 +23,15 @@ gulp.pipe(asciidoctor({
 }));
 ----
 
+.With templates
+[source,javascript]
+----
+var path = require('path')
+
+gulp.pipe(asciidoctor({
+  template_dirs: path.join(__dirname, 'templates')
```
Comment (ggrossetie) [inline doc/documentation.adoc]: :rocket:

File: package.json [hunk 4]
```diff
@@ -1,27 +1,29 @@
 {
   "name": "@asciidoctor/gulp-asciidoctor",
-  "version": "2.1.1",
-  "description": "Gulp AsciiDoctor plugin",
+  "version": "2.2.0",
+  "description": "Gulp Asciidoctor plugin",
```
Comment (ggrossetie) [inline package.json:4]: My pathological obsessiveness for capitalization is thanking you :laughing:

File: doc/documentation.adoc (hunk 2 above)
Comment (ghost) [reply doc/documentation.adoc]: Should standalone = true, header_footer = true  be the default?

File: package.json (hunk 4 above)
Comment (ghost) [reply package.json:4]: I recognized it just a few seconds before 😆

File: package.json (hunk 1 above)
Comment (ghost) [reply package.json]: Sorry, I did not want to offend anybody here. I'll correct it immediately!

File: doc/documentation.adoc (hunk 2 above)
Comment (ggrossetie) [reply doc/documentation.adoc]: The default value in the Asciidoctor CLI is true: https://github.com/asciidoctor/asciidoctor/blob/11aaeabed5935cdb7e65b65eb44b44210f039b87/lib/asciidoctor/cli/options.rb#L15

So I would say yes.

File: package.json (hunk 1 above)
Comment (ggrossetie) [reply package.json]: No worries :smile:

File: doc/documentation.adoc (hunk 2 above)
Comment (ghost) [reply doc/documentation.adoc]: Then it is probably also better to strip header_footer and rely on standalone. I will fix that.

- if only header_footer is set => standalone = value of header_footer, header_footer will be stripped
- if both are set, standalone = value of standalone and header_footer will be stripped

File: index.js [hunk 5]
```diff
@@ -2,52 +2,36 @@
...
-  // Load converter if option is set
-  if (options.converter !== undefined) {
-    var cnv = options.converter
-    if (isClass(options.converter)) {
-      const CnvClass = options.converter
-      cnv = new CnvClass()
-    }
-
-    if (typeof cnv.convert === 'function') {
-      asciidoctor.ConverterFactory.register(cnv, asciidoctorOptions.backend)
```
Comment (ggrossetie) [inline index.js:46]: I think it can still be useful to pass a custom converter. Any particular reason to remove this feature?

Comment (ghost) [reply index.js:46]: No, I don't have any particular reason. I just thought that the templates mechanism will make that obsolete.
I can include it again if you think it is useful

**Comment by ghost** [comment]
@Mogztter I get an error if I try to run the test case with a converter that is exported as a class. I don't know, why that is the case. Do you have any idea?

It says:
convert: undefined method `convert' for #<Proc:0x6b8>
even though I explicitly checked for the existence of the function before.

When I export the class via a function, it works.

**Comment by ggrossetie** [comment]
@henriette-einstein Strange... we do have a test on this feature: https://github.com/asciidoctor/asciidoctor.js/blob/cd07a68127eff5cd080e9926c0a85ff388fcadc8/packages/core/spec/node/asciidoctor.spec.js#L2386 and the class definition is here: https://github.com/asciidoctor/asciidoctor.js/blob/cd07a68127eff5cd080e9926c0a85ff388fcadc8/packages/core/spec/node/asciidoctor.spec.js#L2175-L2202

Can you please share a sample reproduction case? And/or add a failing test?

**Comment by ghost** [comment]
There is a test case included in this branch, that fails when you run "yarn test". It is the same as in the previous release (the master branch). It worked with Asciidoctor 2.1.2.

To reproduce, you can just checkout this branch and run "yarn test" or "npm test". 

I can isolate that test and add some more stuff if that will help. Just give me notice.

**Comment by ghost** [comment]
I just saw, that you are passing in a class. 
In my testcase, I create a new instance using new(). I'll check if that behaves the same as in your case.

**Comment by ghost** [comment]
I tweaked the code a little for debugging. Now I get the following "stacktrace"

     convert: undefined method `convert' for #<Proc:0x6b6>
      at Function.$method_missing (node_modules/asciidoctor-opal-runtime/src/opal.js:3907:56)
      at Function.method_missing_stub (node_modules/asciidoctor-opal-runtime/src/opal.js:1310:35)
      at constructor.$convert (node_modules/@asciidoctor/core/dist/node/asciidoctor.js:8418:35)
      at Function.$convert (node_modules/@asciidoctor/core/dist/node/asciidoctor.js:16304:22)
      at Function.Asciidoctor.convert (node_modules/@asciidoctor/core/dist/node/asciidoctor.js:21693:21)
      at Transform._transform (index.js:86:28)
      at Transform._read (_stream_transform.js:191:10)
      at Transform._write (_stream_transform.js:179:12)
      at doWrite (_stream_writable.js:385:12)
      at writeOrBuffer (_stream_writable.js:367:5)
      at Transform.Writable.write (_stream_writable.js:307:12)
      at Context.<anonymous> (spec/gulp-asciidoctor.spec.js:393:12)
      at processImmediate (internal/timers.js:456:21)

Another info:
I added another testcase (it('should register a custom converter instantiated here',) in which I instantiate the converter in the test case itself (line 406). That works. If I pass the class and instantiate it in index.js, the test fails. (This can be reproduced if you change line 50 in index.js to 

asciidoctor.ConverterFactory.register(new options.converter(), [asciidoctorOptions.backend])

and line 406 of gulp-asciidoctor-spec.js to

      converter: CnvClass

**Comment by ggrossetie** [comment]
@henriette-einstein Thanks I will checkout your branch tomorrow to try to understand why it's failing.

**Comment by ghost** [comment]
I found it! That is probably a pathological edge-case.

What happened is basically

```
const Clazz = MyClass
const myInstance = new Clazz()
asciidoctor.ConverterFactory.register(myInstance, [asciidoctorOptions.backend])
asciidoctorOptions = {
   converter: Clazz
}
asciidoctor.convert(content, asciidoctorOptions)
```

In that case, I have a class that is instantiated and registered before the class itself is passed as an option to the asciidoctor processor. 

I don't know if such a case could be checked in your code. 

I removed the option "converter" after the registration. Then it works.

Sorry for the disturbances, it was my fault.

**Comment by ghost** [comment]
I did not know, that "converter" was actually an option that is used by the underlying Asciidoctor. I therefore changed the name of the plugin option to "cnv" to avoid problems.

**Comment by ggrossetie** [comment]
> I did not know, that "converter" was actually an option that is used by the underlying Asciidoctor. I therefore changed the name of the plugin option to "cnv" to avoid problems.

Yes `converter` is an option but if you do that you will need to pass a "Ruby" class or object. The `asciidoctor.ConverterFactory.register` function will "lift" the JavaScript class or instance to an Opal class/object.

File: doc/documentation.adoc [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+npm install gulp-asciidoctor --save
+----
+
+=== Installation with yarn
+----
+yarn add gulp-asciidoctor 
+----
+
+== Usage
+This plugin can be used to integrate Asciidoctor processing in yout gulp tasks. Here is a typical usage scenario.
```
Comment (ggrossetie) [inline doc/documentation.adoc]:
```suggestion
This plugin can be used to integrate Asciidoctor processing in your gulp tasks. Here is a typical usage scenario:
```

File: doc/documentation.adoc [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+function copyImages(cb) {
+  gulp.src('/path/to/my/files/**/*.jpg')
+    .pipe(gulp.dest('targetdir'))
+  cb()
+}
+
+exports.process = gulp.parallel(processAdocFiles, copyImages)
+----
+
+You can configure the plugin itself by just adding options
```
Comment (ggrossetie) [inline doc/documentation.adoc]:
```suggestion
You can configure the plugin itself by adding options:
```

File: doc/documentation.adoc [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  ...
+
+  .pipe(asciidoctor({
+    extension: '.htm'
+  })
+
+  ...
+----
+
+And you can also configure the underlying Asciidoctor processor via the respective Asciidoctor options
```
Comment (ggrossetie) [inline doc/documentation.adoc]:
```suggestion
And you can also configure the underlying Asciidoctor processor via the respective Asciidoctor options:
```

File: doc/documentation.adoc [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+----
+
+== Options
+The behaviour of the plugin can be controlled via options. There are options that tweak the behaviour of the Gulp-plugin itself and options that controll the underlying Asciidoctor processor.
+
+== Plugin specific Options
+The following options can be used to configure the Gulp-Plugin
+
+extension:: Sets the output extension of the plugin. Defaults to '.html'
+cnv:: Sets a custom converter that should be used for processing. Defaults to undefined.
```
Comment (ggrossetie) [inline doc/documentation.adoc]:
```suggestion
cnv:: Sets a custom converter that should be used for processing. Defaults to `undefined`.
```

File: doc/documentation.adoc [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+== AsciiDoctor Options
+All options of the underlying AsciiDoctor can be set. Please refer to the
```
Comment (ggrossetie) [inline doc/documentation.adoc]:
```suggestion
All options of the underlying Asciidoctor can be set. Please refer to the
```

File: doc/documentation.adoc [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+https://asciidoctor-docs.netlify.app/asciidoctor.js/processor/convert-options[Converter options] page to see, which options are available.
+
+However, the Asciidoctor options
```
Comment (ggrossetie) [inline doc/documentation.adoc]:
```suggestion
However, the following Asciidoctor options:
```

File: doc/documentation.adoc [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+backed:: Sets the converter backend to use. Defaults to 'html5'
+base_dir:: The base directory for resolving relative resources. Defaults to the directory of the currently processed file
+doctype:: The doctype. May be 'article', 'book' or 'inline'. Defaults to 'article'
+safe:: The safemode to use. May be 'unsafe', 'safe', 'server' or 'secure'. Defaults to 'unsafe'
+standalone:: Should headers and footers be included. Defaults to 'true'
+
+== Important
+
+=== Base Directory
+Do not forget to set the AsciiDoctor option `base_dir` if you want to include
```
Comment (ggrossetie) [inline doc/documentation.adoc]:
```suggestion
Do not forget to set the Asciidoctor option `base_dir` if you want to include
```

File: doc/documentation.adoc [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+files from locations that are non-relative to the currently proecssed file.
+
+=== Obsolete option 'header_footer'
+The 'old' option 'header_footer' will be skipped in favor of the new option 'standalone'.
+
+* If 'header_footer' is set and 'standalone' is not set, the processor will receive 'standalone' = value of 'header_footer' option and the option 'header_footer' will be stripped.
+* If both 'header_footer' and 'standalone' are set, the option 'header_footer' will be stripped.
+
+== Change Log
```
Comment (ggrossetie) [inline doc/documentation.adoc]: I think it's one word (not 100% sure):

```suggestion
== Changelog
```

**Comment by ggrossetie** [comment]
@henriette-einstein The code looks fine, I only have minors nitpicks about the documentation, otherwise :+1:

**Comment by ghost** [comment]
@Mogztter I am ready to merge the pull request if that is ok for you

**Comment by ggrossetie** [comment]
@henriette-einstein It looks good to me, I think you can merge it! Well done :+1:

**Comment by ggrossetie** [comment]
:tada: :tada: :tada:

**Pull request merged**